src/adapter/src/rbac.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::catalog::Catalog;
+use crate::command::Command;
+use crate::session::Session;
+use crate::AdapterError;
+use mz_sql::catalog::{CatalogRole, SessionCatalog};
+use mz_sql::plan::Plan;
+use std::error::Error;
+use std::fmt;
+use std::fmt::Formatter;
+
+/// Attributes that allow a role to execute certain plans.
+///
+/// Note: This is a subset of all role attributes used for privilege checks.
+#[derive(Debug)]
+pub enum Attribute {
+    CreateRole,
+    CreateDB,
+    CreateCluster,
+}
+
+impl Attribute {
+    /// Checks if a role has the privilege granted by the attribute.
+    fn check_role(&self, role: &(impl CatalogRole + ?Sized)) -> bool {
```

Comment:
nit: `&(impl CatalogRole + ?Sized)` is interesting, I would kind of prefer to use `&dyn CatalogRole` though because it's more idiomatic. 

Using the `impl CatalogRole` pattern we'd get static dispatch, but the inclusion of `?Sized` and not just `&impl CatalogRole` tells me that we're probably calling this with a `&dyn CatalogRole` already anyways, so I don't think we'd really see a perf improvement from the static dispatch.